README.md
# What is this?
It is a [code assignment](./README-assignment.md) I got in order to show some coding skills.

I've decided to implement it using Javascript.

The assignment scope was limited to a given input format and structure and didn't required further generalizations, so I know that there is a huge room for improvement and extensions (like sorting algorithms, data tied to general timestamps, more alarms and so on).

PRs are welcome.

# Instructions

**Node >= 10 (LTS) is required to run this script.**

### To install: 
1. Clone the repository in a directory `[directory]`
2. `cd [directory]`
3. `npm install`
4. `npm link` (this creates a global symlink to the package)
5. `pond-scan [path to .csv file]`
6. Enjoy.

### To uninstall:
1. Go to repo directory
2. `npm unlink`

OR
1. `npm uninstall -g osmobot-assignment`

### To run tests:
The package includes a **full test coverage**.
Go inside the repo directory and run: 

`npm test`

Once executed you can find the coverage report in a neat html format:
1. `cd coverage`
2. open with a browser the file `index.html`

### Notes:
The code is fully commented.
All the functions and helpers are documented using JSHint conventions. That will trigger autocompletion and code hints in code editors like Visual Studio Code.

CHANGELOG.md
## 1.0.0 (Feb 22, 2019)

Initial release
  - cli.js parse a csv file and gives back the error according to the alarms specified in [README-assignment.md](./README-assignment.md)
  - test suite

test/tests.js
/* eslint-env node, mocha */
/* eslint quotes: ["error", "single"]*/
var ampm = require('../modules/utils').ampm;
var readCsvFile = require('../modules/utils').readCsvFile;
var lowPh = require('../modules/alarm-definitions').lowPh;
var highPh = require('../modules/alarm-definitions').highPh;
var lowDo = require('../modules/alarm-definitions').lowDo;
var changingPh = require('../modules/alarm-definitions').changingPh;
var borderlineDo = require('../modules/alarm-definitions').borderlineDo;
var Alarms = require('../classes/alarm');
var assert = require('assert');

describe('Testing AMPM function', function(){
  it('Tries several inputs', function(){
    assert.equal(ampm('1'), '1am','Error test 1');
    assert.equal(ampm(1), '1am','Error test 1');
    assert.equal(ampm('12'), '12pm', 'Error test 12');
    assert.equal(ampm(12), '12pm', 'Error test 12');
    assert.equal(ampm('13'), '1pm', 'Error test 13');
    assert.equal(ampm(13), '1pm', 'Error test 13');
    assert.equal(ampm('17'), '5pm', 'Error test 17');
    assert.equal(ampm(17), '5pm', 'Error test 17');
    assert.equal(ampm('23.15'), '11pm', 'Error test 23.15');
    assert.equal(ampm(23.15), '11pm', 'Error test 23.15');
    assert.equal(ampm('0'), '12am', 'Error test 0');
    assert.equal(ampm(0), '12am', 'Error test 0');
    assert.equal(ampm('0.15'), '12am', 'Error test 0.15');
    assert.equal(ampm(0.15), '12am', 'Error test 0.15');
    assert.equal(ampm('25'), false, 'Error test 25');
    assert.equal(ampm(25), false, 'Error test 25');
    assert.equal(ampm('hello'), false, 'Error test "hello"');
  });
});

describe('Testing low-ph alarm', function(){
  it('Tries several ph', function(){
    assert.equal(lowPh(2), true, 'Error with ph=2');
    assert.equal(lowPh(9), false, 'Error with ph=9');
    assert.equal(lowPh(6.5), true, 'Error with ph=6.5');
    assert.equal(lowPh(7.9), false, 'Error with ph=7.9');
    assert.equal(lowPh(7.8), false, 'Error with ph=7.8');
    assert.equal(lowPh('2'), true, 'Error with ph="2"');
    assert.equal(lowPh('hello'), false, 'Error with string hello');
  });
});

describe('Testing high-ph alarm', function(){
  it('Tries several ph', function(){
    assert.equal(highPh(2), false, 'Error with ph=2');
    assert.equal(highPh(9), true, 'Error with ph=9');
    assert.equal(highPh(6.5), false, 'Error with ph=6.5');
    assert.equal(highPh(8.6), true, 'Error with ph=8.6');
    assert.equal(highPh(8.5), false, 'Error with ph=8.5');
    assert.equal(highPh('2'), false, 'Error with ph="2"');
    assert.equal(highPh('hello'), false, 'Error with string hello');
  });
});

describe('Testing low-do alarm', function(){
  it('Tries several DOs', function(){
    assert.equal(lowDo(2), true, 'Error with DO=2');
    assert.equal(lowDo(9), false, 'Error with DO=9');
    assert.equal(lowDo(6.5), false, 'Error with DO=6.5');
    assert.equal(lowDo(4), false, 'Error with DO=4');
    assert.equal(lowDo(3.9), true, 'Error with DO=3.9');
    assert.equal(lowDo('2'), true, 'Error with DO="2"');
    assert.equal(lowDo('hello'), false, 'Error with string hello');
  });
});

describe('Testing with changing-ph alarm', function(){
  it('Tries several scenarios', function(){
    assert.equal(changingPh([4,3,2]), true, 'Error scenario 1');
    assert.equal(changingPh([4,3.9,3.8]), false, 'Error scenario 2');
    assert.equal(changingPh([4,3.9,3.2]), false, 'Error scenario 3');
    assert.equal(changingPh([4,4.5,4]), true, 'Error scenario 4');
    assert.equal(changingPh([4,4.2,4]), true, 'Error scenario 5');
    assert.equal(changingPh([4,4.1,4.3]), false, 'Error scenario 6');
    assert.equal(changingPh([4,4.2,4.4]), true, 'Error scenario 7');
    assert.equal(changingPh([4,3]), false, 'Error scenario 8');
    assert.equal(changingPh(['4','3','2']), true, 'Error scenario 9');
    assert.equal(changingPh(['4','3.9',3.8]), false, 'Error scenario 10');
    assert.equal(changingPh(['4','3.9','3.2']), false, 'Error scenario 11');
    assert.equal(changingPh(['4','4.5','4']), true, 'Error scenario 12');
    assert.equal(changingPh(['4','4.2','4']), true, 'Error scenario 13');
    assert.equal(changingPh([4,'4.1','4.3']), false, 'Error scenario 14');
    assert.equal(changingPh([4,4.2,'4.4']), true, 'Error scenario 15');
    assert.equal(changingPh(['4','3']), false, 'Error scenario 16');
    assert.equal(changingPh(['hello', 'bye', 'here']), false, 'Error scenario 17');
    assert.equal(changingPh(4), false, 'Error scenario 18');
  });
});

describe('Testing the borderlineDo alarm', function(){
  it('Tries several scenarios', function(){
    assert.equal(borderlineDo([2,3,4,3]), true, 'Error scenario 1');
    assert.equal(borderlineDo([2,3,5,3]), false, 'Error scenario 2');
    assert.equal(borderlineDo([6,3,4,3]), false, 'Error scenario 3');
    assert.equal(borderlineDo([2,3,4,3,6]), true, 'Error scenario 4');
    assert.equal(borderlineDo([2,3]), false, 'Error scenario 5');
    assert.equal(borderlineDo([2,'3',4,3]), true, 'Error scenario 6');
    assert.equal(borderlineDo(['hello', 'bye', 'here', 'sure']), false, 'Error scenario 7');
    assert.equal(borderlineDo(5), false, 'Error scenario 8');
  });
});

describe('Testing ReadCsvFile utility', function(){
  describe('Tries to read several files', function(){
    this.timeout(20000);
    it('Good file', function(done){
      readCsvFile('./test/test_data.csv', function(err, data){
        if(err){
          done(err);
        } else {
          assert.equal(Array.isArray(data), true, 'Not an array');
          assert.equal(data.length > 0, true, 'Array of zero length');
          done();
        }
      });
    });

    it('File not found', function(done){
      readCsvFile('./test/not_exist', function(err, data){
        assert.equal(err.code, 'ENOENT', 'Wrong error');
        assert.equal(err.message, 'File not found!', 'Wrong message');
        assert.equal(data, null, 'Data is not null');
        done();
      });
    });

    it('Not a CSV, length 0', function(done){
      readCsvFile('./test/empty.csv', function(err, data){
        assert.equal(err.message, 'Not a valid csv: [] length === 0', 'Wrong message');
        assert.equal(data, null, 'Data is not null');
        done();
      });
    });

    it('Not a CSV', function(done){
      readCsvFile('./package.json', function(err, data){
        assert.notEqual(err,null, 'Error is null');
        assert.equal(err.message, 'Not a valid csv: Invalid opening quote at line 2', 'Wrong message');
        assert.equal(data, null, 'Data is not null');
        done();
      });
    });    
  });
});

describe('Tests Alarms class', function(){
  let testData = [];
  before('Preparing test data', function(done){
    this.timeout(20000);
    readCsvFile('./test/test_data.csv', function(err,data){
      if(err){done(err);} else {
        testData = data;
        //console.log(testData);
        done();
      }
    });
  });
  
  it('Test alarm Alarms.ringsLowPh', function(){
    let alarms = new Alarms(testData);
    assert.equal(alarms.ringsLowPh(1), false, 'Alarm triggered @1am');
    assert.equal(alarms.ringsLowPh(3), true, 'Alarm not triggered @ 3am');
    assert.throws(() => {alarms.ringsLowPh(26);}, 'Does not trigger error @26');
  });

  it('Test alarm Alarms.ringsHighPh', function(){
    let alarms = new Alarms(testData);
    assert.equal(alarms.ringsHighPh(1), false, 'Alarm triggered @1am');
    assert.equal(alarms.ringsHighPh(16), true, 'Alarm not triggered @ 16');
    assert.throws(() => {alarms.ringsHighPh(26);}, 'Does not trigger error @26');
  });

  it('Test alarm Alarms.ringsLowDo', function(){
    let alarms = new Alarms(testData);
    assert.equal(alarms.ringsLowDo(1), false, 'Alarm triggered @1am');
    assert.equal(alarms.ringsLowDo(21), true, 'Alarm not triggered @ 21');
    assert.throws(() => {alarms.ringsLowDo(26);}, 'Does not trigger error @26');
  });

  it('Test alarm Alarms.ringsChangingPh', function(){
    let alarms = new Alarms(testData);
    assert.equal(alarms.ringsChangingPh(4), false, 'Alarm triggered @4am');
    assert.equal(alarms.ringsChangingPh(14), true, 'Alarm not triggered @ 14am');
    assert.throws(() => {alarms.ringsChangingPh(26);}, 'Does not trigger error @26');
    assert.equal(alarms.ringsChangingPh(1), false, 'Alarm triggered @1am');
    assert.equal(alarms.ringsChangingPh(0), false, 'Alarm triggered @0am');
  });

  it('Test alarm Alarms.ringsBorderlineDo', function(){
    let alarms = new Alarms(testData);
    assert.equal(alarms.ringsBorderlineDo(3), true, 'Alarm not triggered @3am');
    assert.equal(alarms.ringsBorderlineDo(14), false, 'Alarm triggered @ 14');
    assert.throws(() => {alarms.ringsBorderlineDo(26);}, 'Does not trigger error @26');
    assert.equal(alarms.ringsBorderlineDo(1), false, 'Alarm triggered @1am');
    assert.equal(alarms.ringsBorderlineDo(0), false, 'Alarm triggered @0am');
  });

});

test/cli-test.js
/* eslint-env node, mocha */
/* eslint quotes: ["error", "single"]*/
var ampm = require('../cli').ampm;
var lowPh = require('../cli').lowPh;
var highPh = require('../cli').highPh;
var lowDo = require('../cli').lowDo;
var changingPh = require('../cli').changingPh;
var borderlineDo = require('../cli').borderlineDo;
var assert = require('assert');

describe('Testing AMPM function', function(){
  it('Tries several inputs', function(){
    assert.equal(ampm('1'), '1am','Error test 1');
    assert.equal(ampm(1), '1am','Error test 1');
    assert.equal(ampm('12'), '12pm', 'Error test 12');
    assert.equal(ampm(12), '12pm', 'Error test 12');
    assert.equal(ampm('13'), '1pm', 'Error test 13');
    assert.equal(ampm(13), '1pm', 'Error test 13');
    assert.equal(ampm('17'), '5pm', 'Error test 17');
    assert.equal(ampm(17), '5pm', 'Error test 17');
    assert.equal(ampm('23.15'), '11pm', 'Error test 23.15');
    assert.equal(ampm(23.15), '11pm', 'Error test 23.15');
    assert.equal(ampm('0'), '12am', 'Error test 0');
    assert.equal(ampm(0), '12am', 'Error test 0');
    assert.equal(ampm('0.15'), '12am', 'Error test 0.15');
    assert.equal(ampm(0.15), '12am', 'Error test 0.15');
    assert.equal(ampm('25'), false, 'Error test 25');
    assert.equal(ampm(25), false, 'Error test 25');
    assert.equal(ampm('hello'), false, 'Error test "hello"');
  });
});

describe('Testing low-ph alarm', function(){
  it('Tries several ph', function(){
    assert.equal(lowPh(2), true, 'Error with ph=2');
    assert.equal(lowPh(9), false, 'Error with ph=9');
    assert.equal(lowPh(6.5), true, 'Error with ph=6.5');
    assert.equal(lowPh(7.9), false, 'Error with ph=7.9');
    assert.equal(lowPh(7.8), false, 'Error with ph=7.8');
    assert.equal(lowPh('2'), true, 'Error with ph="2"');
    assert.equal(lowPh('hello'), false, 'Error with string hello');
  });
});

describe('Testing high-ph alarm', function(){
  it('Tries several ph', function(){
    assert.equal(highPh(2), false, 'Error with ph=2');
    assert.equal(highPh(9), true, 'Error with ph=9');
    assert.equal(highPh(6.5), false, 'Error with ph=6.5');
    assert.equal(highPh(8.6), true, 'Error with ph=8.6');
    assert.equal(highPh(8.5), false, 'Error with ph=8.5');
    assert.equal(highPh('2'), false, 'Error with ph="2"');
    assert.equal(highPh('hello'), false, 'Error with string hello');
  });
});

describe('Testing low-do alarm', function(){
  it('Tries several DOs', function(){
    assert.equal(lowDo(2), true, 'Error with DO=2');
    assert.equal(lowDo(9), false, 'Error with DO=9');
    assert.equal(lowDo(6.5), false, 'Error with DO=6.5');
    assert.equal(lowDo(4), false, 'Error with DO=4');
    assert.equal(lowDo(3.9), true, 'Error with DO=3.9');
    assert.equal(lowDo('2'), true, 'Error with DO="2"');
    assert.equal(lowDo('hello'), false, 'Error with string hello');
  });
});

describe('Testing with changing-ph alarm', function(){
  it('Tries several scenarios', function(){
    assert.equal(changingPh([4,3,2]), true, 'Error scenario 1');
    assert.equal(changingPh([4,3.9,3.8]), false, 'Error scenario 2');
    assert.equal(changingPh([4,3.9,3.2]), false, 'Error scenario 3');
    assert.equal(changingPh([4,4.5,4]), true, 'Error scenario 4');
    assert.equal(changingPh([4,4.2,4]), true, 'Error scenario 5');
    assert.equal(changingPh([4,4.1,4.3]), false, 'Error scenario 6');
    assert.equal(changingPh([4,4.2,4.4]), true, 'Error scenario 7');
    assert.equal(changingPh([4,3]), false, 'Error scenario 8');
    assert.equal(changingPh(['4','3','2']), true, 'Error scenario 9');
    assert.equal(changingPh(['4','3.9',3.8]), false, 'Error scenario 10');
    assert.equal(changingPh(['4','3.9','3.2']), false, 'Error scenario 11');
    assert.equal(changingPh(['4','4.5','4']), true, 'Error scenario 12');
    assert.equal(changingPh(['4','4.2','4']), true, 'Error scenario 13');
    assert.equal(changingPh([4,'4.1','4.3']), false, 'Error scenario 14');
    assert.equal(changingPh([4,4.2,'4.4']), true, 'Error scenario 15');
    assert.equal(changingPh(['4','3']), false, 'Error scenario 16');
    assert.equal(changingPh(['hello', 'bye', 'here']), false, 'Error scenario 17');
    assert.equal(changingPh(4), false, 'Error scenario 18');
  });
});

describe('Testing the borderlineDo alarm', function(){
  it('Tries several scenarios', function(){
    assert.equal(borderlineDo([2,3,4,3]), true, 'Error scenario 1');
    assert.equal(borderlineDo([2,3,5,3]), false, 'Error scenario 2');
    assert.equal(borderlineDo([6,3,4,3]), false, 'Error scenario 3');
    assert.equal(borderlineDo([2,3,4,3,6]), true, 'Error scenario 4');
    assert.equal(borderlineDo([2,3]), false, 'Error scenario 5');
    assert.equal(borderlineDo([2,'3',4,3]), true, 'Error scenario 6');
    assert.equal(borderlineDo(['hello', 'bye', 'here', 'sure']), false, 'Error scenario 7');
    assert.equal(borderlineDo(5), false, 'Error scenario 8');
  });
});

cli.js
#!/usr/bin/env node

//Import classes and utils
var readCsvFile = require('./modules/utils').readCsvFile;
var ampm = require('./modules/utils').ampm;
var Alarms = require('./classes/alarm');    //Import the Alarms class

//Read the command line arguments
const [,, ...args] = process.argv;


/**
 *Executes at startup
 *
 */
function main(){
  //Checks if arguments have been supplied
  if(!args[0]){
    console.log('Please specify a .csv file');
  } else {
    //Reads the csv file
    readCsvFile(args[0], function(err, parsedArray){
      if(err){
        console.error(err);
        return err;
      } else {
        //Instantiate the Alarms class
        var alarms = new Alarms(parsedArray);
        //Scans all the 24 hours for alarms
        for (let i = 0; i < 24; i++) {
          if(alarms.ringsLowPh(i)){
            console.log(ampm(i) + ': ALERT low-pH');
          }
          if(alarms.ringsHighPh(i)){
            console.log(ampm(i) + ': ALERT high-pH');
          }
          if(alarms.ringsLowDo(i)){
            console.log(ampm(i) + ': ALERT low-DO');
          }
          if(alarms.ringsChangingPh(i)){
            console.log(ampm(i) + ': ALERT changing-pH');
          }
          if(alarms.ringsBorderlineDo(i)){
            console.log(ampm(i) + ': ALERT borderline-DO');
          }                       
        }
      }
    });
  }
}

//Begins execution
main();

classes/alarm.js
#!/usr/bin/env node

var lowPh = require('../modules/alarm-definitions').lowPh;
var highPh = require('../modules/alarm-definitions').highPh;
var lowDo = require('../modules/alarm-definitions').lowDo;
var changingPh = require('../modules/alarm-definitions').changingPh;
var borderlineDo = require('../modules/alarm-definitions').borderlineDo;

/**
 *Evaluates alarms
 *
 * @class Alarms
 */
class Alarms {
  /**
   *Creates an instance of Alarms.
   * @param {[[String]]} parsedCsv The CSV with daily datas
   * @memberof Alarms
   */
  constructor(parsedCsv){
    this.dataMatrix = parsedCsv;
  }

  /**
   *Check the lowPH condition at a given hour
   *
   * @param {number} hour24 The hours when the alarm is checked
   * @returns {boolean} True if the alarm rings
   * @memberof Alarms
   */
  ringsLowPh(hour24){
    //Tests if an out of range parameter has been submitted
    if(hour24 < 0 || hour24 > 23){
      let err = new Error('Wrong parameter (hour24 has to be between 0 and 23)');
      err.code = 'ERRPARAM';
      throw err;
    } else {
      return lowPh(this.dataMatrix[hour24][2]);
    }
  }

  /**
   *Check the highPh condition at a given hour
   *
   * @param {number} hour24 The hours when the alarm is checked
   * @returns {boolean} True if the alarm rings
   * @memberof Alarms
   */
  ringsHighPh(hour24){
    //Tests if an out of range parameter has been submitted
    if(hour24 < 0 || hour24 > 23){
      let err = new Error('Wrong parameter (hour24 has to be between 0 and 23)');
      err.code = 'ERRPARAM';
      throw err;
    } else {
      return highPh(this.dataMatrix[hour24][2]);
    }
  }

  /**
   *Check the lowDO condition at a given hour
   *
   * @param {number} hour24 The hours when the alarm is checked
   * @returns {boolean} True if the alarm rings
   * @memberof Alarms
   */
  ringsLowDo(hour24){
    //Tests if an out of range parameter has been submitted
    if(hour24 < 0 || hour24 > 23){
      let err = new Error('Wrong parameter (hour24 has to be between 0 and 23)');
      err.code = 'ERRPARAM';
      throw err;
    } else {
      return lowDo(this.dataMatrix[hour24][1]);
    }
  }

  /**
   *Check the changingPH condition at a given hour
   *
   * @param {number} hour24 The hours when the alarm is checked
   * @returns {boolean} True if the alarm rings
   * @memberof Alarms
   */
  ringsChangingPh(hour24){
    //Tests if an out of range parameter has been submitted
    if(hour24 < 0 || hour24 > 23){
      let err = new Error('Wrong parameter (hour24 has to be between 0 and 23)');
      err.code = 'ERRPARAM';
      throw err;
    } else {
      //Builds up the measurement needed for changingPh alarm status evaluation
      let measurementArray = [];
      for (let i = 0; i < 3; i++) {
        //Avoids to get non-existent data
        if(hour24 - i >= 0){
          measurementArray.push(this.dataMatrix[hour24-i][2]);
        }        
      }
      return changingPh(measurementArray);
    }
  }

  /**
   *Check the borderlineDO condition at a given hour
   *
   * @param {number} hour24 The hours when the alarm is checked
   * @returns {boolean} True if the alarm rings
   * @memberof Alarms
   */
  ringsBorderlineDo(hour24){
    //Tests if an out of range parameter has been submitted
    if(hour24 < 0 || hour24 > 23){
      let err = new Error('Wrong parameter (hour24 has to be between 0 and 23)');
      err.code = 'ERRPARAM';
      throw err;
    } else {
      //Builds up the measurement needed for changingPh alarm status evaluation
      let measurementArray = [];
      for (let i = 0; i < 4; i++) {
        //Avoids to get non-existent data
        if(hour24 - i >= 0){
          measurementArray.push(this.dataMatrix[hour24-i][1]);
        }        
      }
      //console.log(measurementArray);
      return borderlineDo(measurementArray);
    }
  }
}

module.exports = Alarms;

modules/alarm-definitions.js
#!/usr/bin/env node

//Threshold values for alarms
const HIGH_PH = 8.5;
const LOW_PH = 7.8;
const CHANGING_PH = 0.2;
const LOW_DO = 4;
const BORDERLINE_DO = 5;


/**
 *Returns true if the ph is under a given threshold value
 *
 * @param {string | number} ph Ph value
 * @returns {boolean} True if ph is low
 */
function lowPh(ph){
  //Check if ph is a number or a string representing number
  if(isNaN(ph)){
    //Not a number/stringified number. Return false (no alarm)
    return false;
  } else {
    //Check threshold value
    if(parseFloat(ph) < LOW_PH){
      //low Ph
      return true;
    } else {
      //Ph is ok
      return false;
    }
  }
}

/**
 *Returns true if the ph is above a given threshold value
 *
 * @param {string | number} ph Ph value
 * @returns {boolean} True if ph is high
 */
function highPh(ph){
  //Check if ph is a number or a string representing number
  if(isNaN(ph)){
    //Not a number/stringified number. Return false (no alarm)
    return false;
  } else {
    //Check threshold value
    if(parseFloat(ph) > HIGH_PH){
      //low Ph
      return true;
    } else {
      //Ph is ok
      return false;
    }
  }
}

/**
 *Returns true if the DO is under a given threshold value
 *
 * @param {string | number} doValue DO value
 * @returns {boolean} True if DO is low
 */
function lowDo(doValue){
  //Check if doValue is a number or a string representing number
  if(isNaN(doValue)){
    //Not a number/stringified number. Return false (no alarm)
    return false;
  } else {
    //Check threshold value
    if(parseFloat(doValue) < LOW_DO){
      //low DO
      return true;
    } else {
      //DO is ok
      return false;
    }
  }
}

/**
 * Returns true if the ph change rate is over a given threshold value in the last 2 hours
 *
 * @param {[string | number]} phArray The last 3 hourly measurement of pH, in chronological order
 * @returns {boolean} True if ph is changing too quickly
 */
function changingPh(phArray){
  //Checks that phArray is an array
  if(!Array.isArray(phArray)){
    //Not an array. 
    return false;
  } else {
    //Check if there are enough data to check the changingPh condition (at least 3 measures in the last 2 hours)
    if (phArray.length < 3) {
      //Not enough data
      return false;
    } else {
      //Checks that there are not common strings - like "hello" - in the array.
      for (let i = 0; i < phArray.length; i++) {
        const element = phArray[i];
        if(isNaN(element)){
          return false;
        }
      }
      //Check the changingPh condition
      if(Math.abs(parseFloat(phArray[0])-parseFloat(phArray[1])) >= CHANGING_PH && Math.abs(parseFloat(phArray[1])-parseFloat(phArray[2])) >= CHANGING_PH){
        return true;
      } else {
        return false;
      }
    }
  }
}

/**
 *  Returns true if the DO is under a given threshold value in the last 3 hours
 *
 * @param {[string | number]} doArray The last 4 hourly measurement of DO, in chronological order
 * @returns {boolean} True if DO is consistently low under a certain threshold
 */
function borderlineDo(doArray){
  //Checks that phArray is an array
  if(!Array.isArray(doArray)){
    //Not an array
    return false;
  } else {
    //Check if there are enough data (at least 4 measurements in the last 3 hours)
    if (doArray.length < 4){
      //Not enough data
      return false;
    } else {
      let wentAboveLimit = false; //Initial value
      for (let i = 0; i < 4; i++) {
        const element = parseFloat(doArray[i]);
        //Checks that there are not common strings - like "hello" - in the array.
        if(isNaN(element)){
          return false;
        }
        //Checks if anytime during the [for] cycle the DO level was above the threshold BORDERLINE_DO
        if(!wentAboveLimit){
          if(element < BORDERLINE_DO){
            wentAboveLimit = false;
          } else {
            //DO is above the threshold limit. 
            wentAboveLimit = true;
          }
        }
      }
      //The alarm triggers if all 4 measures are under the threshold BORDERLINE_DO (wentAboveLimit === false)
      if(wentAboveLimit){
        return false;
      } else {
        return true;
      }
    }
  }
}

module.exports.lowPh = lowPh;
module.exports.highPh = highPh;
module.exports.lowDo = lowDo;
module.exports.changingPh = changingPh;
module.exports.borderlineDo = borderlineDo;

modules/utils.js
#!/usr/bin/env node

var parse = require('csv-parse');
const fs = require('fs');

/**
 *Converts 24h to AMPM notation
 *
 * @param {string | number} hour24 The hour in 24h format
 * @returns {string | boolean}  A string of the hour in am/pm format or false is hour24 is not a valid hour
 */
function ampm(hour24){
  if (isNaN(hour24)){
    return false;
  } else {
    var hour = parseFloat(hour24);
    if (hour >= 0 && hour < 1){
      return '12am';
    } else if (hour >= 1 && hour < 12 ) {
      return Math.floor(hour).toString() + 'am';
    } else if (hour >= 12 && hour < 13){
      return '12pm';
    } else if (hour >= 13 && hour < 24){
      return (Math.floor(hour)-12).toString() + 'pm';
    } else {
      return false;
    }
  }
}

/**
 *Read a CSV file
 *
 * @param {string} path The path of csv file
 * @param {function(Error, [[string]])} cb Callback(err, data)
 */
function readCsvFile(path, cb){
  //Reads the csv file
  fs.readFile(path, 'utf8', function(err, data){
    if(err){
      let readFileError = new Error('Error on reading file: ' + err.message);
      readFileError.returnedError = err;
      readFileError.code = err.code;
      if(err.code === 'ENOENT') {
        //console.error('File not found!');
        readFileError.message = 'File not found!';        
      }
      return cb(readFileError, null);
    } else {
      //Parse the csv starting from line 2 to avoid headers
      parse(data,{from_line: 2},function(err,parsedArray){
        if(err){
          //console.error('Not a valid csv');
          let invalidCsv = new Error('Not a valid csv: ' + err.message);
          invalidCsv.returnedError = err;
          return cb(invalidCsv, null);
        } else {
          if(parsedArray.length === 0){
            //console.error('Not a valid csv');
            let errorLength = new Error('Not a valid csv: [] length === 0');
            return cb(errorLength, null);
          } else {
            return cb(null, parsedArray);
          }
        }
      });
    }
  });
}

//Exporting statements
module.exports.ampm = ampm;
module.exports.readCsvFile = readCsvFile;